sql/src/main/java/io/crate/expression/scalar/string/TrimFunction.java

```diff
@@ -0,0 +1,243 @@
+                    return new OneCharTrimFunction(
+                        new FunctionInfo(
+                            new FunctionIdent(NAME, ImmutableList.of(datatypes.get(0))),
+                            datatypes.get(0)),
+                        ' '
+                    );
+                } else if (datatypes.size() == 2) {
+                    return new TwoArgsTrimFunction(
+                        new FunctionInfo(new FunctionIdent(NAME, datatypes), datatypes.get(0))
+                    );
+                } else {
+                    return new ThreeArgsTrimFunction(
+                        new FunctionInfo(new FunctionIdent(NAME, datatypes), datatypes.get(0))
+                    );
+                }
+            }
+        });
+    }
+
+    private abstract static class TrimBaseFunction extends Scalar<String, String> {
+
+        private final FunctionInfo info;
+
+        TrimBaseFunction(FunctionInfo info) {
+            this.info = info;
+        }
+
+        @Override
+        public FunctionInfo info() {
+            return info;
+        }
+
+        HashSet<Character> getUniqueChars(String charsToTrim) {
+            return new HashSet<>(Chars.asList(charsToTrim.toCharArray()));
+        }
+    }
+
+    private static class ThreeArgsTrimFunction extends TrimBaseFunction {
+
+        ThreeArgsTrimFunction(FunctionInfo info) {
+            super(info);
+        }
+
+        @Override
+        public Scalar<String, String> compile(List<Symbol> arguments) {
+            assert arguments.size() == 3 : "number of args must be 3";
+
+            Symbol modeSymbol = arguments.get(2);
+            if (!Literal.isLiteral(modeSymbol, DataTypes.STRING)) {
+                return this;
+            }
+            TrimMode mode = TrimMode.of((String) ((Input) modeSymbol).value());
+            if (mode != TrimMode.BOTH) {
+                return this;
+            }
+
+            Symbol charsToTrimSymbol = arguments.get(1);
+            if (!Literal.isLiteral(charsToTrimSymbol, DataTypes.STRING)) {
+                return this;
+            }
+
+            String charsToTrim = (String) ((Input) charsToTrimSymbol).value();
+            if (charsToTrim.length() == 1) {
+                return new OneCharTrimFunction(info(), charsToTrim.charAt(0));
+            }
+            return this;
+        }
+
+        @Override
+        public String evaluate(TransactionContext txnCtx, Input<String>[] args) {
+            assert args.length == 3 : "number of args must be 3";
+
+            String target = args[0].value();
+            if (target == null) {
+                return null;
+            }
+            String charsToTrimArg = args[1].value();
+            if (charsToTrimArg == null) {
+                return target;
+            }
+            TrimMode mode = TrimMode.of(args[2].value());
+
+            HashSet<Character> charsToTrim = getUniqueChars(charsToTrimArg);
+
+            int start = 0;
```

Comment:
Might want to consider having something like `int start = mode.getStartIdx(charsToTrim)` and `int len = mode.getTrimmedLength(charsToTrim) `.

I think this would allow us to get rid of the two/three args distinction without adding any significant overhead.

Reply:
The second method needs to know about the start index and vice-versa, otherwise, pointers will cross. Also, both methods have to take the target string and probably we would need to have reloaded version of those methods for a single char. So, I am not sure whether it can be an improvement if I get the suggestion right.

Reply:
It would be mostly about removing the distinction of `if (mode.isTrailingOrBoth()) {` by using polymorphism.